feat: libsignal wasm

jlucaso1 opened the pull request:
Replaces the JS libsignal with the Rust core compiled to WebAssembly, keeping the auth state on disk exactly as it is today.

## What changed

Session operations run as pure functions: the caller reads a snapshot, the protocol runs in memory, and the mutations come back as an explicit changeset applied in one write. Nothing calls back into JS mid-operation, so an operation cannot re-enter the caller's transaction or take locks in a conflicting order. That re-entrancy is what produced `message with old counter` failures under load.

The identity row is locked alongside the session, because applying a changeset writes both and the group path does a read-modify-write on that same row.

## Storage is unchanged

Sessions and group sender keys are written in the shape a pre-WASM release reads, not as the bridge's bytes. Rolling back is swapping the package: no conversion step, no migration. Identity keys keep the same 33-byte prefixed form, and pre-keys are untouched.

This costs about 0.3 ms per message against storing the native bytes. Serialization is worth its own change, with the breaking note that deserves, rather than extra surface area here.

One caveat: a record the v1 model cannot express keeps its bridge bytes. Reads accept both shapes, so that costs compatibility for one row instead of failing the write. Reaching it takes genuine out-of-order delivery.

## Outbound counters

The core leases 64 outbound counters ahead of durability, and the legacy record has nowhere to carry that ceiling, so the export materialized the reservation and the whole batch burned on every operation. Sends landed on the wire at counters 0, 64, 128, 192, and the peer derived 63 skipped keys for each. After 20 group messages the stored row was 199,724 bytes and a decrypt cost 3,488 us; a direct-message row passed the skipped-key limit around the 32nd message and fell back to bridge bytes, losing the rollback guarantee above.

This branch waives the lease, which the core exposes as a per-record declaration (oxidezap/whatsapp-rust#1211). That is sound because the changeset is persisted before the ciphertext reaches the wire, and it is the guarantee pre-WASM releases gave. A direct-message row after 40 messages is now 919 bytes and still legacy.

## Derivation cache

A sender-key state memoizes two curve derivations, but the memo lives in the record, which we rebuild per operation. Keeping the record alive instead would go stale the moment anything rotates the key. The bridge keys on the signing key's public bytes, which has no such window, and hands the result back through the core's prewarm setters (oxidezap/whatsapp-rust#1213), which verify it belongs to the state.

`encryptGroupMessage` goes from 71.8 to 62.0 us.

## Numbers

`wabench` against the same server, three runs per side, interleaved, median reported. Intel Core Ultra 9 275HX, 24 threads, `--allocator stable --cpu-policy auto`. Certificate verification is neutralised in both benchmark worktrees, since the test server cannot sign with WhatsApp's certificate key.

Direct messages, 3000 at 300/s:

| | develop | this branch |
|---|---|---|
| p50 | 2.62 ms | **0.59 ms** |
| p99 | 6.82 ms | **2.62 ms** |
| client CPU | 0.908 cores | **0.488 cores** |
| peak RSS | 293 MB | **200 MB** |
| sustained rate | 294.1 /s | **299.9 /s** |

Only this branch holds the offered 300/s; `develop` settles below it and accumulates lag.

Group messages, 500 at 50/s, which is what `develop` sustains:

| | develop | this branch |
|---|---|---|
| p50 | 10.49 ms | **1.57 ms** |
| p99 | 41.94 ms | **6.29 ms** |
| client CPU | 0.817 cores | **0.143 cores** |
| peak RSS | 303 MB | **176 MB** |

Past that rate `develop` stops following: offered 100/s it delivers 53.4/s at a 2,013 ms median. This branch holds 300/s with a 1.05 ms median.

Group receive is not covered by either scenario, and it is where the counter fix lands hardest: a microbenchmark of consecutive decrypts goes from 3,488 us on the twentieth message to a flat 82 us, with the stored row dropping from 199,724 to 441 bytes.

## Fixes found along the way

- Identity records were keyed by the bare user while sessions use the full address, so the transaction layer locked a different row than the one being written. This is what let an encrypt and a decrypt run concurrently on one session.
- `getSenderKeyDistributionMessage` seeded an empty record before asking the builder to create one, which loads as a state-less session and fails. Distributing your own sender key was broken, so no group could be sent to.
- Device 99 never migrated: discovery probed one address shape and the code re-derived the other, and the destination was built as a plain LID jid, which device 99 is rejected for.
- A consumed pre-key was reported by the core but never deleted.
- A pre-key delete failure discarded an already-decrypted message.
- `Array::from(undefined)` threw across the boundary on a legacy sender-key state stored without `senderMessageKeys`.
- Buffers in a stored sender key can be a byte array or base64 text depending on whether the row went through `BufferJSON`. Only the array was accepted, so a store that round-trips through it read every key as absent and surfaced as `protobuf encoding was invalid`.
- Sender-key states were written newest first, while the JS backend reads the last entry as current and drops the first on overflow, so a record with more than one state was upside down.

## Cleanup

`SessionCipher` and `SessionBuilder` are gone, along with the adapter's session, identity and pre-key stores that existed to serve them, and the legacy JSON session fallback, which was unreachable and stored a message key's seed as its cipher key. The `SignalStorage` interface goes from twelve members to two, and the SIMD wasm drops 7.6% to 959,313 bytes.

## Compatibility

The rc.9 fixture holds real sessions and ciphertexts produced by baileys@7.0.0-rc.9, with no Rust in the pipeline. The suite decrypts them, keeps the conversation going, and checks stored records are untouched when only read. The three wire-parity suites in the bridge now replay recorded rc.9 vectors instead of importing it.

## Validation

583 tests in the Baileys package, 168 in the bridge plus 21 Rust unit tests under `wasm-pack`, lint clean.

The bridge tracks the Rust core at the commit that merged the prewarm setters, so it has to be built from source or published before this lands.

jlucaso1 committed:
feat: extend signalStorage type to include loadSignedPreKey method

jlucaso1 committed:
feat: group support

jlucaso1 committed:
refactor: replace libsignal dependency with whatsapp-rust-bridge functions

jlucaso1 committed:
fix: import Long type for improved type handling

jlucaso1 committed:
chore: update whatsapp-rust-bridge to version 0.4.0-alpha.2

Salientekill commented:
I'm following along and testing it out 🫡🙂

jlucaso1 committed:
fix: update whatsapp-rust-bridge to version 0.4.0-alpha.3 (migration session and sender keys)

jlucaso1 commented:
Good news:

The PR is ready for production testing (no need to delete sessions or sender keys anymore, this will be auto migrated, only in case reverting this is necessary)

4relial commented:
I've tested this in a production environment. There are several cases where the bot suddenly stops receiving upsert messages, and it only starts working again after rescan. I'm still monitoring it, but for now that's the main issue I've encountered.

jlucaso1 committed:
fix: update whatsapp-rust-bridge to version 0.4.0-alpha.4

purpshell commented:
Would love if the logging architecture was implemented the same way

purpshell commented:
> > It was merged but the purpshell reverted.
> 
> we need to handle it differently.

Related to this @Salientekill (even though this is off topic) check out rc9

YonkoSam commented:
will start testing this today

jlucaso1 committed:
chore: bump whatsapp-rust-bridge

jlucaso1 committed:
fix: simplify address creation in group message handling

jlucaso1 committed:
chore: update whatsapp-rust-bridge to version 0.4.1

jlucaso1 commented:
> Works perfectly, but when an error happens, the rust library returns a string instead of a node error. The error before the TypeError is from handling offline messages btw.
> 
> ```js
> 2025-11-25 09:39:52 PM - Baileys [error]: transaction failed, rolling back
>     error: SessionCipher.decryptWhisperMessage failed: SessionNotFound(ProtocolAddress { name: "*****************_1", device_id: DeviceId(0) })
> 2025-11-25 09:39:52 PM - Baileys [error]: failed to decrypt message
>     key: {
>         "addressingMode": "lid",
>         "fromMe": false,
>         "id": "ACAB6948EEE4FF9573A586EA5835B289",
>         "participant": "***********@lid",
>         "participantAlt": "**********@s.whatsapp.net",
>         "remoteJid": "120363321350778834@g.us"
>     }
>     err: SessionCipher.decryptWhisperMessage failed: SessionNotFound(ProtocolAddress { name: "***********_1", device_id: DeviceId(0) })
>     messageType: msg
>     sender: 120363321350778834@g.us
>     author: *************@lid
>     isSessionRecordError: false
> > 2025-11-25 09:39:52 PM - Baileys [error]: error in handling message - Cannot read properties of undefined (reading 'toString')
> > TypeError: Cannot read properties of undefined (reading 'toString')
> >    at decrypt (file:///home/rin/neru-wa/node_modules/baileys/lib/Utils/decode-wa-message.js:264:74)
> >    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
> >    at async file:///home/rin/neru-wa/node_modules/baileys/lib/Socket/messages-recv.js:988:17
> ...
> ```

Nice catch, I'll check this case. Thank you again

jlucaso1 committed:
fix: improve error handling in message decryption

lotte25 commented:
Sorry for not reporting earlier. After restarting the bot, it ***very*** slowly starts to handle offline messages, does nothing (marks delivered after hours and doesn't emit to message upsert) and after that, it just keeps trying to... i don't know. Send a message or send ACKs?
To test this, i left the bot on the entire day.

<img width="205" height="581" alt="image" src="https://github.com/user-attachments/assets/885f1029-f288-40eb-aff1-08babe9b9fba" />

Event buffer times out sometimes. Don't know if it's part of the issue.
```js
2025-11-28 01:52:11 AM - Baileys [warn]: Buffer timeout reached, auto-flushing
2025-11-28 01:52:11 AM - Baileys [debug]: Flushing event buffer
    bufferCount: 1
 ```
 
 These logs appear without end:
```js
2025-11-27 04:40:56 PM - Baileys [debug]: Using PN identity for @s.whatsapp.net conversation
    to: ***********@s.whatsapp.net
    ownId: ***********:32@s.whatsapp.net
2025-11-27 04:40:56 PM - Baileys [debug]: assertSessions call with jids
    jids: [
        "***********@s.whatsapp.net"
    ]
2025-11-27 04:40:56 PM - Baileys [debug]: adding device identity
    jid: ***********@s.whatsapp.net
2025-11-27 04:40:56 PM - Baileys [debug]: sending message to 1 devices
    msgId: 3EB0350E757ED75F6AC4B6
2025-11-27 04:40:56 PM - Baileys [debug]: Added message to retry cache: ***********@s.whatsapp.net/3EB0350E757ED75F6AC4B6
2025-11-27 04:40:56 PM - Baileys [debug]: sendRetryRequest: requested placeholder resend (3EB0350E757ED75F6AC4B6) for message AC99876403844AC1BC09B746D19F31E3 (scheduled)
2025-11-27 04:40:57 PM - Baileys [debug]: Using PN identity for @s.whatsapp.net conversation
    to: ***********@s.whatsapp.net
    ownId: ***********:32@s.whatsapp.net
2025-11-27 04:40:57 PM - Baileys [debug]: assertSessions call with jids
    jids: [
        "***********@s.whatsapp.net"
    ]
2025-11-27 04:40:57 PM - Baileys [debug]: adding device identity
    jid: ***********@s.whatsapp.net
2025-11-27 04:40:57 PM - Baileys [debug]: sending message to 1 devices
    msgId: 3EB05854BAE5C0A44CBF46
2025-11-27 04:40:57 PM - Baileys [debug]: Added message to retry cache: ***********@s.whatsapp.net/3EB05854BAE5C0A44CBF46
2025-11-27 04:40:57 PM - Baileys [debug]: sendRetryRequest: requested placeholder resend (3EB05854BAE5C0A44CBF46) for message ACEEFACA99E950046BFA099F6B1C50A7 (scheduled)
2025-11-27 04:40:57 PM - Baileys [debug]: Using PN identity for @s.whatsapp.net conversation
    to: ***********@s.whatsapp.net
    ownId: ***********:32@s.whatsapp.net
2025-11-27 04:40:57 PM - Baileys [debug]: assertSessions call with jids
    jids: [
        "56928941133@s.whatsapp.net"
    ]
2025-11-27 04:40:57 PM - Baileys [debug]: adding device identity
    jid: 56928941133@s.whatsapp.net
2025-11-27 04:40:57 PM - Baileys [debug]: sending message to 1 devices
    msgId: 3EB03E8ACE081222D118B6
2025-11-27 04:40:57 PM - Baileys [debug]: Added message to retry cache: 56928941133@s.whatsapp.net/3EB03E8ACE081222D118B6
2025-11-27 04:40:57 PM - Baileys [debug]: sendRetryRequest: requested placeholder resend (3EB03E8ACE081222D118B6) for message ACAC0BD95449E31685CC79F4F005EC09 (scheduled)
```

If baileys (finally) tries to decode a message, the no sender key record comes up.
```js
2025-11-28 01:41:47 AM - Baileys [error]: transaction failed, rolling back
    error: no sender key state: no sender key record for group 120363036771877682@g.us sender  *********_1::42
2025-11-28 01:41:47 AM - Baileys [error]: failed to decrypt message
    key: {
        "addressingMode": "lid",
        "fromMe": false,
        "id": "3EB08A3CEFE85D58C6AA84",
        "participant": " *********:42@lid",
        "participantAlt": " *********@s.whatsapp.net",
        "remoteJid": "120363036771877682@g.us"
    }
    err: no sender key state: no sender key record for group 120363036771877682@g.us sender  *********_1::42
    messageType: skmsg
    sender: 120363036771877682@g.us
    author:  *********:42@lid
    isSessionRecordError: false
```

jlucaso1 committed:
chore: update whatsapp-rust-bridge to version 0.4.4

jlucaso1 committed:
chore: update whatsapp-rust-bridge to version 0.4.5

jlucaso1 committed:
chore: update whatsapp-rust-bridge to version 0.4.6 and set logger in libsignal

jlucaso1 committed:
chore: update whatsapp-rust-bridge to version 0.4.7

jlucaso1 committed:
chore: update whatsapp-rust-bridge to version 0.4.8

Salientekill commented:
When the coex is added, please let me know so I can test it.

jlucaso1 committed:
fix: force 99 deviceId to be hosted device

jlucaso1 committed:
fix: update whatsapp-rust-bridge to version 0.5.0-alpha.1

jlucaso1 committed:
perf: binary serialization works better with lru-cache

Salientekill commented:
I can't wait for it to be implemented.

jlucaso1 committed:
Merge branch 'master' of https://github.com/WhiskeySockets/Baileys into feat-libsignal-wasm

jlucaso1 committed:
Merge branch 'master' into feat-libsignal-wasm

Resolve conflicts between the libsignal->WASM migration and master:

- package.json: drop native `libsignal` dependency (replaced by
  whatsapp-rust-bridge); keep bridge at master's 0.5.4.
- yarn.lock: regenerated, removing libsignal + curve25519-js entries.
- Utils/generics.ts: take master's baileysVersion [2,3000,1035194821];
  drop duplicate `import type Long`.
- Utils/auth-utils.ts: drop unused `NodeCache` import (cache replaced by
  LRUCache); keep `Boom` import used by master's auth assertion.
- Signal/libsignal.ts encryptGroupMessage: adopt master's shared
  `ensureSenderKeyAndCreateSkdm` helper + transaction-scoped atomicity,
  using the WASM GroupCipher(storage, group, addr) 3-arg API.
- Signal/libsignal.ts ensureSenderKeyAndCreateSkdm: adapt master's helper
  to the WASM SignalStorage API (storeSenderKey(string, Uint8Array));
  import SenderKeyRecord from whatsapp-rust-bridge.

jlucaso1 committed:
chore: merge develop into feat-libsignal-wasm

Rebases the WASM Signal backend onto the pnpm monorepo. Conflict
resolution notes:

- Signal/Group/* and its regression test are deleted: the Rust bridge
  now owns the group ratchet. SenderKeyName's storage key is unchanged
  (`group::user::device` on both sides), so persisted sender keys stay
  readable.
- libsignal.ts was reconstructed on top of develop's version rather than
  taking either side of the textual merge, so the Stage 0-10 concurrency
  work survives intact: every transactWith scope, the pre-resolved
  wireJid locking and the pinned-resolver TOCTOU fix are preserved and
  only the backend calls were swapped.
- getSessionInfo used to cast loadSession's result to an object with
  getOpenSession(). The bridge returns the raw serialized record, so the
  cast type-checked while the method returned null at runtime, silently
  disabling the retry path's registration-id and base-key collision
  checks. It now decodes the record with WAProto's RecordStructure, and
  a new test drives a real bridge session end to end to pin that format.
- validateSession keeps its haveOpenSession() check by deserializing the
  record, instead of dropping it.
- auth-utils keeps the NodeCache -> LRUCache swap, applied to develop's
  rewritten makeCacheableSignalKeyStore.

jlucaso1 committed:
fix: scope device domainType per device, guard SKDM input

Addresses review feedback on #2067.

extractDeviceJids hoisted `domainType` to the user result, so the first
hosted device (or device 99) rewrote the domain for every device that
followed it in the same list — a plain PN device listed after a hosted
one was addressed as @hosted. Scope it per device; regression test
added, and it fails without the fix.

Also: guard the sender key distribution message instead of asserting it
non-null, drop two `any` casts in favour of real types, and correct the
migratedSessionCache TTL comment (3 days, not 7).

jlucaso1 committed:
fix(signal): honor signature results and keep pre-WASM sessions readable

Addresses the backend-migration review on #2067.

Curve.verify accepted every well-formed signature. The JS libsignal threw
on a bad signature, so `try { verify(); return true }` was correct there;
the bridge instead RETURNS false and only throws on malformed input, so
the result was being discarded. Tampered account signatures and Noise
certificate signatures both passed. Return the bridge's verdict.

Sessions written before this branch are the JS libsignal object
(`SessionRecord.serialize()` returned `{_sessions, version}`, never
bytes). Two paths mishandled them:

- migrateSession ran them through SessionRecord.deserialize, which
  deliberately flattens a legacy object to an EMPTY record. The PN → LID
  copy was therefore skipped and the session stranded behind the now-LID
  lookup. Move the record across verbatim instead.
- loadSession handed the whole record to the bridge, whose converter
  takes `Object.keys(_sessions)[0]`. That is insertion-ordered, so after
  a ratchet rotation it is a CLOSED state and the upgraded session is
  unusable. Select the open state (the converter also accepts a bare
  entry) and pass that.

getSessionInfo and validateSession read the legacy shape directly, so the
retry protections cover unconverted sessions too.

Also normalize the bridge's `DuplicatedMessage(...)` to
MISSING_KEYS_ERROR_TEXT — messages-recv matches that text exactly to ACK
a replayed stanza (487), so duplicates were driving retry/resend loops —
and keep device 99 on HOSTED_LID when the user is already hosted-LID,
instead of downgrading it into the PN namespace.

Known gap: states beyond the open one are still dropped, because the
published bridge writes `previous_sessions: []`. Preserving the backlog
needs the bridge-side converter.

jlucaso1 committed:
fix(signal): never revive a closed legacy session, keep live LID sessions

Follow-up review on #2067.

loadSession fell back to handing the bridge the whole legacy record when
no state was open. The bridge then promotes `_sessions`' first key to the
current session, so a closed state came back to life and encryption would
run under a ratchet the peer had already dropped. Report "no session"
instead and let it renegotiate.

migrateSession copied a legacy PN record onto the LID key unconditionally.
Before the legacy branch existed such records were skipped, which
preserved a post-upgrade LID session by accident; with a real copy the
check has to be explicit, so read the destination rows (already inside the
lock scope) and leave a live LID session alone.

Assert the duplicate-decrypt message by equality rather than substring:
messages-recv compares it with `===`, so a prefix would break the 487 ACK
while still passing toThrow. Also fix a comment claiming HOSTED is 1.

jlucaso1 committed:
feat(signal): convert pre-WASM sessions through the core's typed model

The storage adapter's own legacy converter copies the libsignal message key
in as a cipher key and zeroes the mac/iv. That material is a SEED the core
re-derives the split from, so every session it converted failed its MAC on
the first ciphertext the old build had enciphered.

Port the core's typed interop boundary (importLegacySessionRecordV1 /
projectLegacySessionRecordV1, behind the wacore legacy-session-interop
feature) into the 0.5.x bridge and route loadSession through it. That also
picks the OPEN state rather than `_sessions`' first key, which after a
rotation is a closed one, and preserves the archived states the ad-hoc path
dropped.

The conversion is bidirectional, so the migration is no longer one-way: a
projected session is readable by the pre-WASM build again.

Evidence, from a session generated by baileys@7.0.0-rc.9 with no Rust in the
pipeline (fixtures/legacy-session-rc9.json):
- decrypts DM and group ciphertexts the JS backend left pending;
- keeps the conversation going in both directions afterwards;
- leaves the stored record untouched on read-only calls;
- round-trips import -> project -> import to identical bytes;
- the real rc.9 picks the projection back up and keeps sending — proving a
  rollback needs no re-pairing.

Bridge side also updates wacore to the current main (waproto moved from
prost to buffa) and fixes the resulting type errors.

jlucaso1 committed:
fix(bridge): key identity records by the full address

save_identity, is_trusted_identity and get_identity used address.name(), which
drops the device, while sessions use the full address. The transaction layer
then locked a different id than the record being written, so an encrypt and a
decrypt on one session could run at the same time and rewind the sending
chain. The peer rejects the next ciphertext as "message with old counter".

jlucaso1 committed:
fix(bridge): delete the pre-key a prekey message consumed

The core reports the consumed id instead of removing the record, so this path
left a spent one-time pre-key in storage where it could be handed out again.

jlucaso1 committed:
feat(bridge): run signal operations over a snapshot

The callback store calls back into JS several times per operation. That forces
the caller to predict which records an operation will touch so it can lock them
up front, and it lets a nested scope take those records in an order that
inverts against another scope's.

Here the caller passes in everything the operation can need, the protocol runs
in memory, and the mutations come back as an explicit changeset: one read, one
write, no callbacks in between. Effects the core reports rather than applies,
like the consumed pre-key, are part of that changeset.

jlucaso1 committed:
feat(signal): move the session path onto the snapshot API

decryptMessage, encryptMessage and injectE2ESession now read a snapshot, run
the operation, and apply the returned changes in a single write under one lock.
A prekey message names the key it consumes, so that record is read up front
instead of being fetched mid-operation.

This drops the pinned-resolution storage wrapper and the identity parsing that
existed to work around re-entrancy, and turns rejections into Errors, since the
bridge rejects with plain strings and callers read .message.

jlucaso1 committed:
fix(signal): address device 99 and live LID sessions during migration

Migration probed only `user.<device>` and then re-derived the address from the
jid, which for device 99 resolves to the hosted form. A row written under
either shape was therefore never matched, and building the destination through
transferDevice produced a plain LID jid, which device 99 is rejected for. Probe
both shapes, carry the address the row was found under, and send device 99 to
the hosted LID domain.

The bridge-bytes branch also copied over the destination without checking it,
so a session established on the LID key after the PN one was overwritten. It
now takes the same guard as the legacy branch.

jlucaso1 committed:
fix(bridge): review fixes for the snapshot path

- key the address cache by name AND device, since two devices of one user share
  a name and the cache would hand back the wrong address
- keep a pre-key removal reported by the store when the core names none
- log, rather than raise, a failure to delete a consumed pre-key: the message is
  already decrypted and dropping it would lose it to a cleanup problem
- refuse save_pre_key and save_signed_pre_key on the snapshot store, which has
  no way to report those writes back to the caller
- validate the output path in the rollback fixture, and run cargo fmt

jlucaso1 committed:
ci: build the bridge from source when a change touches it

pnpm install copies the last PUBLISHED bridge tarball into the workspace, so a
change to the Rust crate is invisible to every other package until it ships. A
PR that adds an export fails with "does not provide an export named ...", which
says nothing about the code under review and cannot be fixed without publishing
first.

Move the toolchain setup into a composite action that compiles the crate when
the pull request touches it, and keep the prebuilt otherwise. bridge-build.yml
now shares the same action instead of repeating the pinned Binaryen setup.

jlucaso1 committed:
style(bridge): drop redundant to_string in format args

Clippy only reaches this file now that the crate is compiled in CI.

jlucaso1 committed:
fix(signal): lock the identity row with the session, and pick one device-99 source

Applying a changeset writes identity-key as well as session, but only the
session was locked. signalStorage.saveIdentity does a read-modify-write on that
same row while holding its lock, and clears the session when the key differs,
so the two paths could overwrite each other. transactWith acquires refs in a
fixed order, so naming both is safe.

Probing both device-99 address shapes could also find two rows that share one
destination: migrating both deleted both and kept whichever landed last. Take
the hosted row, which is the shape written today, and leave the other in place.

jlucaso1 committed:
ci: force the bridge build in the job that exists to compile it

bridge-build.yml also runs for lockfile and workflow changes, where the
detector correctly reports the bridge as untouched. That left the job without
wasm-pack while still running the build, so it now asks for the source build
explicitly. Changes to the action itself trigger the workflow too, since it is
the only job that exercises them.

jlucaso1 committed:
fix(signal): prefer the open row when a device has two address shapes

Taking the hosted one unconditionally stranded the live session whenever the
hosted row was the closed one.

jlucaso1 committed:
perf(bridge): report the peer identity only when it changes

The core reasserts trust on first use during every encrypt and decrypt, and
the store reported that as a change each time, so the caller rewrote the
identity row on every message, once per device. Report it only when the key is
new or different, which is what the callback path already did.

Adds coverage for the identity-change path, where the old session is voided and
a new one is built in the same operation, and pins the stored identity to the
33-byte prefixed form a pre-WASM release also writes.

jlucaso1 committed:
test(signal): generate the one-time pre-keys the fixtures use

Borrowing a pair off a throwaway initAuthCreds() suggested a relationship to
the peer that does not exist. A one-time pre-key is just a fresh pair, so make
that explicit rather than reusing the peer's signed pre-key, which would put
one key in two roles.

jlucaso1 committed:
ci: read the changed files without a pipe

grep -q exits at the first match, so on a pull request large enough to fill the
pipe the writer takes SIGPIPE and pipefail fails the condition, selecting the
prebuilt on exactly the changes that need the source build.

jlucaso1 committed:
fix(bridge): treat an absent JS property as absent

Reflect::get answers Ok(undefined) for a property that is not there, so
get_object never returned None. A legacy sender-key state written without
senderMessageKeys, which is how the JS libsignal stored a state with none, then
reached Array::from(undefined) and threw across the boundary, killing the group
operation instead of migrating the row.

The ok_or_else callers were dead for the same reason: a missing senderChainKey
fell through as undefined instead of raising.

jlucaso1 committed:
test(bridge): unit-test the Rust side

The changeset semantics and the JS property reads were only reachable through
the TS suite, which exercises them end to end and cannot pin the edges: an
unchanged identity reporting nothing, a replaced one voiding the session, a
write the changeset cannot carry being refused, and an absent property reading
as absent rather than as undefined.

Runs under wasm-pack test --node, since the crate only builds for wasm32. The
modules alias #[test] to wasm_bindgen_test to keep the annotations plain.

jlucaso1 committed:
test(bridge): skip the Rust tests when wasm-pack is absent

AGENTS.md tells contributors they do not need the Rust toolchain, since the
postinstall pulls the prebuilt. Putting test:rust in the default chain broke
`pnpm test` for exactly those checkouts with "wasm-pack: command not found".

The gate is an if/else rather than `&& ... ||`, so a failing Rust test still
fails the run instead of falling through to the skip message. CI keeps calling
test:rust directly, where the toolchain is always installed.

jlucaso1 committed:
test(bridge): detect wasm-pack from Node instead of the shell

The previous gate used `command -v` inside an if/then/fi, which is POSIX only:
on Windows pnpm runs scripts through cmd.exe, and PowerShell rejects it with a
parse error, so `pnpm test` was broken for those contributors instead of
skipping.

The lookup now walks PATH in Node, applying PATHEXT on Windows, and spawns the
binary directly. An npm-installed wasm-pack is a .cmd, which Node will not
spawn without a shell, so that case passes one quoted command string.

jlucaso1 committed:
Merge branch 'develop' into feat-libsignal-wasm

The three fixes from #2751 exist here in a different form, so the overlaps were
resolved rather than taken from one side:

- crypto.ts keeps the bridge's verifySignature, which is what this branch calls
- signal.ts keeps the device-99 hosted rule on top of the per-device domain fix
- auth-utils.ts keeps the LRUCache work with the escaped lock separator, so the
  file stays text for git
- curve-verify.test.ts takes develop's copy; both sides cover the same cases and
  keeping theirs avoids diverging from the base again

jlucaso1 committed:
fix: review follow-ups after the develop merge

- build.yml checks out without credentials, matching the other workflows
- toTypedRecord skips a hole in _sessions instead of casting undefined, which
  would throw and lose the live sessions alongside the broken entry
- the legacy Buffer envelope is typed as number[] | string and decoded on the
  runtime type rather than assuming base64
- the plain-address device-99 test asserts the destination holds the bytes
- the lock test records event order instead of asserting on a delay, which on a
  slow machine would also hold when the encrypt simply had not run yet
- the sender-key shapes share one table

identity-format keeps handing the bundle a prefixed key: the bridge normalises
the identity inside a snapshot but decodes bundle.identityKey strictly, so a
bare 32 bytes is rejected there.

jlucaso1 committed:
fix(signal): build a sender key without seeding an empty record

getSenderKeyDistributionMessage stored an empty SenderKeyRecord before asking
the builder to create one. The builder handles a missing row by creating a
record, but a stored empty one loads as a state-less session and the create
then fails with InvalidSenderKeySession, so distributing your own sender key
was broken and no group could be sent to. The JS backend passed the record
object rather than its bytes and did not have that distinction.

Nothing covered getSenderKeyDistributionMessage, which is why it went
unnoticed; group-sender-key.test.ts now does.

rollback-step1 becomes a test instead of a script that never ran. It pins what
a rollback can and cannot carry: a session this backend advanced projects back
to the legacy shape and survives the round trip, while a sender key written
here cannot be parsed by the old build and has to be redistributed.

jlucaso1 committed:
chore(bridge): pin the core that persists skipped-key seeds

MessageKey gained a seed field, and the migration builds that struct with an
exhaustive literal, so the bump is a compile error until every site states what
it writes. These two sites carry no seed: they read a legacy record whose
skipped keys were already stored as bare material.

jlucaso1 committed:
feat(signal): store sessions in the shape a pre-WASM release reads

The session row is written as the legacy JSON record instead of the bridge's
bytes, so this branch does not change what sits on disk and going back to an
older release is swapping the package rather than converting storage.

A record the v1 model cannot express keeps its bridge bytes: reads already
accept both shapes, so the fallback costs compatibility for that one row
instead of failing the write. It also covers a gap between the two limits, since
the engine prunes skipped keys above MAX_MESSAGE_KEYS plus a threshold while the
v1 import refuses anything above the limit itself.

Costs about 0.3 ms per message and 0.13 cores at 300 msg/s against storing the
native bytes, and still runs roughly 3x faster than the JS backend it replaces.

jlucaso1 committed:
test(signal): say what the assertions actually check

Two comments still described byte comparisons that the legacy storage shape
replaced with a base-key check.

The sender-key assertion also had to assert the row exists first: a missing one
would make Buffer.from throw on its own, and the toThrow below would pass
having proved nothing about the stored shape.

jlucaso1 committed:
fix(signal): rebuild a session the bridge cannot read

A stored row that fails to decode was raised on read, so every retry hit the
same record and the conversation stayed stuck until someone deleted it by hand.
The JS backend this replaces caught the deserialization failure and treated the
session as absent.

A prekey message carries everything needed to build a replacement, so that path
retries without the row. A whisper message still fails, since there is nothing
to rebuild from and starting a session that decrypts nothing would be worse.
The retry only runs after a decode failure, so the happy path is unchanged.

The sender-key assertion now checks the row really holds bytes: a legacy object
would make Buffer.from throw on its own and pass the toThrow below.

jlucaso1 committed:
test(signal): prove the rebuilt session is persisted

The recovery test only checked that decrypt returned plaintext, so a path that
rebuilt the session and dropped the changeset would have passed while the next
message hit the same bad row.

It now asserts the stored record was replaced with a live one, that the pre-key
the message consumed is gone, and that the recovered session carries a reply on
its own.

jlucaso1 committed:
feat(bridge): project a sender-key record into the v1 shape

Every field of a sender-key state has a v1 counterpart, so unlike a session
this conversion loses nothing: the group format already persists seeds rather
than derived material. The record is exposed as a function instead of being
applied inside store_sender_key, so the bridge keeps writing native bytes and
the caller decides what shape its storage keeps.

Reading a legacy record also stopped carrying an empty private signing key as
present. The JS backend wrote an empty Buffer for every sender key received
from someone else, and Some(empty) fails validation the moment anything reads
the record's components.

jlucaso1 committed:
feat(signal): store group sender keys in the pre-WASM shape

The last row that still diverged from what an older release reads. With this,
rolling back keeps groups working instead of needing every sender key
redistributed, verified by running a group through this branch and then handing
the storage to develop: it read the pending message, kept sending, and adopted a
new key without any conversion step.

Also drops the empty record written before processing a distribution message.
The builder creates one when the row is absent, and that pre-write was the only
remaining path storing a sender key outside the policy above.

jlucaso1 committed:
fix(bridge): keep the empty private signing key in the projected record

The JS backend wrote senderSigningKey.private as an empty Buffer for every
sender key received from someone else, so omitting the field made the row
distinguishable from one it produced. It reads either shape, but the point of
this projection is to be indistinguishable, not merely readable.

The rollback suite's header still claimed sender keys had no projection and
could not be read by the old build, which the tests below it now disprove.

jlucaso1 committed:
build(bridge): add a profiling build that keeps symbols

A CPU profile of the release artifact shows wasm-function[N]: the release
profile strips, and wasm-opt drops the name section. wasm-pack's own
--profiling still builds the release profile, so `pnpm build:profile` drives
cargo and wasm-bindgen directly, skips wasm-opt, and resolves the wasm-bindgen
matching the crate's schema from wasm-pack's cache.

The release path is untouched: same flags, same output size.

jlucaso1 committed:
build(bridge): fail the profiling build with an actionable message

Falling through to a global wasm-bindgen is worse than stopping: it is almost
never the version the crate was linked against, and the error it produces talks
about schema numbers instead of what to do. It now names the version and the
two ways to get it.

The wasm-pack metadata profile went with it. That path drives cargo directly,
so the section was never read.

jlucaso1 committed:
fix(bridge): read legacy sender keys whose buffers are base64 text

A stored sender key is legacy JSON, and the buffers inside it can be
written two ways: as a byte array by plain JSON.stringify, or as base64
text by Baileys' own BufferJSON.replacer. The decoder only accepted the
array, so a store that round-trips its rows through BufferJSON read every
key as absent.

Absent then defaulted to empty, which built a record with no chain seed
and no signing key. That record fails to deserialize, and the failure
surfaced as "protobuf encoding was invalid" with nothing pointing at the
row it came from.

Accept both shapes, and stop defaulting the fields the record cannot be
built without so a bad row names itself instead of failing later.

jlucaso1 committed:
fix(bridge): write legacy sender-key states oldest first

The core keeps the current state at the front of the record and prunes
from the back. The legacy shape is the mirror of that: the JS backend
reads the last entry as the current state and drops the first one when
the record overflows.

Both conversions carried the core order through unchanged, so a record
with more than one state was written upside down. Rolling back to a
pre-WASM release would send under a stale key, and the sixth
distribution message would evict the freshest state instead of the
oldest.

Reverse on the way out and on the way back in.

jlucaso1 committed:
fix(bridge): resolve wasm-bindgen from PATH in profiling builds

The error told the reader to install the exact wasm-bindgen-cli and put
it on PATH, but the lookup only scanned the wasm-pack cache. Following
that advice on a clean machine hit the same throw, leaving build:profile
unusable until an unrelated normal build populated the cache.

Scan PATH after the cache, with the same exact-version check, so a
mismatched global build still cannot slip in.

jlucaso1 committed:
test(bridge): stop forcing a GC between benchmark iterations

mitata already defaults to collecting once before a benchmark; asking for
"inner" collects between every iteration instead. That penalises the WASM
side out of proportion, because each collection also drains the
FinalizationRegistry that frees wasm-bindgen handles, and the cost lands
inside the measurement.

The numbers were not just noisy, they pointed the wrong way: encodeNode
read as 11.45x slower than the JS encoder and measures 2.24x faster
without the forced collection, and decodeNode goes from 11.98x slower to
2.56x faster.

jlucaso1 committed:
fix(bridge): waive the core's counter lease

The core leases 64 outbound counters ahead of durability so the send path
only needs a flush once per batch. Carrying that ceiling forward means
persisting it, and the record shape we write has nowhere to put it: the
component export materializes the reservation, so the whole batch burned
on every operation instead of once per batch.

Consecutive sends landed on the wire at counters 0, 64, 128, 192, and the
peer derived 63 skipped keys for each one. After 20 group messages the
stored row was 199,724 bytes and a decrypt cost 3,488 us against 97 us
for the first; a DM row passed MAX_LEGACY_MESSAGE_KEYS around the 32nd
message and fell back to bridge bytes, which is the rollback guarantee
going away.

We hand the changeset back to the caller, which persists it before the
ciphertext reaches the wire, so there is nothing left for the lease to
protect. Every record goes through counter_lease.rs rather than calling
the core directly: a path that forgets goes back to leasing without
failing, and the symptom shows up somewhere else entirely.

Group decrypt is now flat at ~82 us with a 441 byte row, and a DM row
after 40 messages is 919 bytes and still legacy.

jlucaso1 committed:
fix(bridge): waive the counter lease on the store paths too

Records were waived where they are loaded, but a record the protocol
builds for itself never passes a load: the store path is the only place
that sees it before it is serialized and cached. Both stores kept the
default lease on such a record, so the policy read differently depending
on where a record came from.

No such record can reach a send today, because the builders and the
ciphers hold separate adapters and therefore separate caches. Verified:
five sends through one reused cipher, over a record the group builder
created, leave the iteration at 5 either way. That is a property of the
call sites though, not of the policy, and the whole point of routing
records through one place is that a path which forgets fails silently.

Also pin the shape in the base64 sender-key test. It rewrites the row
with BufferJSON.replacer, which rewrites a { type: 'Buffer', data: [...] }
object as well as a real Buffer, so the assertion now states that rather
than leaving a reader to work it out.

jlucaso1 committed:
perf(bridge): keep the sender-key derivation the record cannot keep

A sender-key state memoizes the signing key's basepoint multiplication
and the verifier's Edwards entries, but the memo lives in the record and
we rebuild the record from storage on every operation. Every group
message repaid a derivation whose inputs never changed.

Keeping the record alive instead is the obvious fix and the wrong one:
it would go stale the moment anything rotated the key, and encrypting
under a chain the peer already dropped is worse than encrypting slowly.
Keying on the signing key's public bytes has no such window. A rotated
key is a different entry by construction, so a stale entry is
unreachable rather than wrong, and the core's prewarm setters verify
that what they are handed belongs to the state before installing it.

Measured on the group path, cipher rebuilt per message against one kept
warm: the gap closes from 14.8 to 1.9 us on encrypt and from 10.2 to 5.9
on decrypt. End to end, encryptGroupMessage goes from 71.8 to 62.0 us.

Needs oxidezap/whatsapp-rust#1213, which added the setters.

jlucaso1 committed:
refactor(bridge): drop the legacy session JSON fallback

The adapter would migrate a pre-WASM session object field by field when
the store handed it one instead of bytes. Baileys never hands it one:
`loadSession` converts a legacy record through the core's typed model
first, or reports no session, so the branch was unreachable from the
only consumer.

It was also wrong. The fallback stored a message key's seed as its
cipher key and zeroed the mac and iv, so the first ciphertext the old
build enciphered failed its MAC. That is why the conversion moved to the
typed model in the first place, and the comment at the call site has
said so since.

Its test asserted that the session opened and that encrypt returned a
WhisperMessage, never that the peer could read it, so it would not have
caught any of that. Removing both leaves one way to read a legacy
record, the one that works.

jlucaso1 committed:
refactor(bridge): drop the callback-based session API

The snapshot calls replaced SessionCipher and SessionBuilder for the
whole direct-message path, and Baileys is the only consumer, so nothing
reached them any more. Removing the pair took the adapter's session,
identity, pre-key and signed-pre-key stores with it: those existed to
serve those two types. The adapter now backs the group path alone, which
needs sender keys and nothing else.

The cascade the compiler then found came out too: five payload structs,
the address and identity caches, the raw-store probe, and the extern
declarations for eleven JS callbacks the bridge no longer calls. The
SignalStorage interface goes from twelve members to two.

wasm-opt output drops 7.6%, from 1,038,376 to 959,313 bytes for the SIMD
build and 1,095,158 to 1,012,934 for the other.

Three test files went with the API. Their scenarios did not: the bundle
round trip now runs through processBundleWithSnapshot, and simultaneous
initiation, injection followed by an incoming prekey message, and the
guarantee that reprocessing a bundle does not discard a live session are
expressed against the snapshot calls in snapshot_api.test.ts. What is
left of storage_adapter.test.ts is the one case that still applies: a
rejected write must not be cached as if it had succeeded.

jlucaso1 committed:
test: shrink the legacy fixture and record the wire parity vectors

The rc.9 fixture was 2,270 lines of pretty-printed JSON, most of it
whitespace around base64. Minified and stripped of the four fields no
test reads, it is one line and 10,878 bytes, down from 27,637.

The three wire parity suites no longer import baileys@7.0.0-rc.9 at all.
They call the same functions through test/helpers/legacy-wire.ts, which
serves what rc.9 produced for each node from a recorded file. Parity is
still checked against the JS implementation, just not by running it: a
node with no recorded vector throws rather than passing quietly, and the
header says how to re-record.

Attribute order is part of the wire format, so the vector key preserves
key order. Sorting it collapsed the two nodes in the order-sensitivity
test onto one entry, which passed for the wrong reason until the replay
run caught it.

crypto-parity now compares against node's own crypto: MD5 and HKDF are
standard, so the reference is the algorithm, not another implementation.

appstate-parity and noise-session still need the dependency, because
LT_HASH_ANTI_TAMPERING and makeNoiseHandler are client logic rather than
primitives, and porting 1,600 lines of them is its own change.

Also widen signalStorage's return type: the bridge's SignalStorage went
from twelve members to two, and this file still calls the rest itself.

jlucaso1 committed:
fix(bridge): make the recorded wire vectors typecheck and round-trip

Three problems with the vector harness, one of which broke CI.

`decodeBinaryNode` returned `unknown`, so every caller reading `.tag` or
`.attrs` failed the test typecheck. Jest never caught it because ts-jest
does not typecheck; the CI step that runs tsc did.

The buffer handling was inert. `Buffer.toJSON` runs before any replacer,
so the `__b` branch was unreachable and the revive side looked for a
shape that is never produced. Nothing broke yet because no recorded node
carries binary content, but the first one that did would have come back
as a plain object instead of a Buffer. Store and revive the shape that
actually appears.

The merge on write is read-modify-write, so recording has to be
single-threaded. Say so where the instructions are, and write through a
temp file so an interrupted run cannot leave truncated JSON behind.

Also rename crypto-parity's helpers to nodeMd5/nodeHkdf and say
node:crypto in the titles: the reference stopped being the JS client
when the comparison moved to the standard algorithms.

jlucaso1 committed:
test(bridge): reject a malformed Buffer in a recorded vector

The revive guard matched any object with type "Buffer" and an array
under data. Buffer.from truncates out-of-range values rather than
rejecting them, so a hand-edited or truncated vector would have revived
as wrong bytes and the comparison would have run against them.

Check the byte range and the exact key set, and throw when something
claims to be a Buffer but is not one, so the failure names itself.

jlucaso1 committed:
chore(bridge): drop five unused dependencies

hashify, hmac, prost, serde-wasm-bindgen and simd-adler32 are declared
but never referenced. prost in particular is not a leftover from the
core: the core does not depend on it at all, so it entered the lockfile
purely through this declaration.

Two that cargo-machete also flagged stay, and now say why. getrandom is
never called; it is there to turn on the wasm_js backend, without which
the wasm32 target refuses to build. rand's sys_rng is what make_rng
falls back to for seeding, and while another crate happens to enable it
today, key generation must not depend on that staying true.

Nothing changes in the artifact: dead code elimination was already
dropping these, so the wasm is byte-identical at 959,313. The lockfile
loses 26 lines and two crates.

jlucaso1 committed:
fix(signal): read a pre-v1 legacy session record

libsignal's deserialize runs a v1 migration that copies a record-level
registrationId onto every entry missing one, and haveOpenSession then
requires the entry to have it. An auth state written before that
migration still keeps the id only at the top.

We required it on the entry, so for such a record every entry looked
unusable: hasOpenLegacySession returned false, readSessionBytes reported
no session, and pending whisper messages failed with SessionNotFound
after the upgrade. The session was live the whole time.

Fall back to the record-level id in both the open-session check and the
conversion, which is the same migration libsignal does.

jlucaso1 committed:
fix(signal): satisfy prettier, and drop the libsignal-node mention

The ternary in entryRegistrationId needed parentheses. AGENTS.md still
described Signal/ as wrapping libsignal-node, which this branch replaced
with the bridge.

jlucaso1 committed:
test(bridge): handle the Result from add_sender_key_state

The only warning left in the crate, in a test helper that dropped the
Result on the floor. Now the build is clean under clippy with -D
warnings across every target and feature.

jlucaso1 committed:
test(bridge): port the signal bench to the snapshot calls

Dropping SessionCipher and SessionBuilder left this bench importing
them, so `pnpm bench` died on the second file. It now holds the snapshot
itself and applies each changeset, which is what a consumer does; the
wrapper keeps the old method names so the benchmark bodies are
unchanged.

Still 7.5x, 9.6x and 4.0x faster than libsignal-node on encrypt,
decrypt and round trip.

jlucaso1 committed:
fix(bridge): clear the bench session when the peer's identity is replaced

`SnapshotCipher` models what the Baileys store does with a changeset, and
it ignored `sessionCleared`. The core sets that when a peer's identity key
is replaced: the ratchet built on the old key is void, so it returns no new
session and asks the caller to drop the row instead. Applying only
`changes.session` left the cipher holding bytes the core had disowned, and
every later operation would have run on them.

Not reachable from the benchmark as written, which never rotates an
identity, but the point of this class is to mirror `applyChanges`, and it
did not. The production path in `Signal/libsignal.ts` was already correct.
